server/backends/migration_cache/migration_cache.go

```diff
@@ -0,0 +1,184 @@
+
+	srcCache, err := getCacheFromConfig(env, *cacheMigrationConfig.Src)
+	if err != nil {
+		return err
+	}
+	destCache, err := getCacheFromConfig(env, *cacheMigrationConfig.Dest)
+	if err != nil {
+		return err
+	}
+	mc := &MigrationCache{
+		Src:  srcCache,
+		Dest: destCache,
+	}
+
+	if env.GetCache() != nil {
+		log.Warningf("Overriding configured cache with migration_cache. If running a migration, all cache configs" +
+			" should be nested under the cache.migration block.")
+	}
+	env.SetCache(mc)
+
+	return nil
+}
+
+func validateCacheConfig(config CacheConfig) error {
+	numConfigs := 0
+	v := reflect.ValueOf(config)
+	for i := 0; i < v.NumField(); i++ {
+		if !v.Field(i).IsNil() {
+			numConfigs++
+		}
+	}
+
+	if numConfigs != 1 {
+		return status.FailedPreconditionError("exactly one config must be set")
+	}
+
+	return nil
+}
+
+func getCacheFromConfig(env environment.Env, cfg CacheConfig) (interfaces.Cache, error) {
+	err := validateCacheConfig(cfg)
+	if err != nil {
+		return nil, status.FailedPreconditionErrorf("error validating migration cache config: %s", err)
+	}
+
+	if cfg.DiskConfig != nil {
+		c, err := diskCacheFromConfig(env, cfg.DiskConfig)
+		if err != nil {
+			return nil, err
+		}
+		return c, nil
+	} else if cfg.PebbleConfig != nil {
+		c, err := pebbleCacheFromConfig(env, cfg.PebbleConfig)
+		if err != nil {
+			return nil, err
+		}
+		return c, nil
+	}
+
+	return nil, status.FailedPreconditionErrorf("error getting cache from migration config: no valid cache types")
+}
+
+func diskCacheFromConfig(env environment.Env, cfg *DiskCacheConfig) (*disk_cache.DiskCache, error) {
+	opts := &disk_cache.Options{
+		RootDirectory:     cfg.RootDirectory,
+		Partitions:        cfg.Partitions,
+		PartitionMappings: cfg.PartitionMappings,
+		UseV2Layout:       cfg.UseV2Layout,
+	}
+	return disk_cache.NewDiskCache(env, opts, cache_config.MaxSizeBytes())
+}
+
+func pebbleCacheFromConfig(env environment.Env, cfg *PebbleCacheConfig) (*pebble_cache.PebbleCache, error) {
+	opts := &pebble_cache.Options{
+		RootDirectory:          cfg.RootDirectory,
+		Partitions:             cfg.Partitions,
+		PartitionMappings:      cfg.PartitionMappings,
+		MaxSizeBytes:           cfg.MaxSizeBytes,
+		BlockCacheSizeBytes:    cfg.BlockCacheSizeBytes,
+		MaxInlineFileSizeBytes: cfg.MaxInlineFileSizeBytes,
+		AtimeUpdateThreshold:   cfg.AtimeUpdateThreshold,
+		AtimeWriteBatchSize:    cfg.AtimeWriteBatchSize,
+		AtimeBufferSize:        cfg.AtimeBufferSize,
+		MinEvictionAge:         cfg.MinEvictionAge,
+	}
+	c, err := pebble_cache.NewPebbleCache(env, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	c.Start()
+	env.GetHealthChecker().RegisterShutdownFunction(func(ctx context.Context) error {
+		return c.Stop()
+	})
+	return c, nil
+}
+
+func (mc MigrationCache) WithIsolation(ctx context.Context, cacheType interfaces.CacheType, remoteInstanceName string) (interfaces.Cache, error) {
+	//TODO implement me
+	panic("implement me")
```

Comment:
nit: I know this is placeholder code but typically we don't use panics in server code ever.

You could instead

```
return nil, status.UnimplementedError("not yet implemented")
```

Reply:
Done